src/proxy/http/h2.rs

```diff
@@ -0,0 +1,164 @@
+use std::{fmt, error::Error as StdError};
+use std::marker::PhantomData;
+
+use futures::{Future, Poll};
+use http;
+use hyper::{body::Payload, client::conn::{self, Handshake, SendRequest}};
+
+use super::{Body, Error};
+use svc;
+use task::{ArcExecutor, BoxSendFuture, Executor};
+use transport::connect;
+
+#[derive(Debug)]
+pub struct Connect<C, B> {
+    connect: C,
+    executor: ArcExecutor,
+    _marker: PhantomData<fn() -> B>,
+}
+
+#[derive(Debug)]
+pub struct Connection<B> {
+    tx: SendRequest<B>,
+}
+
+pub struct ConnectFuture<C: connect::Connect, B> {
+    executor: ArcExecutor,
+    state: ConnectState<C, B>,
+}
+
+enum ConnectState<C: connect::Connect, B> {
+    Connect(C::Future),
+    Handshake(Handshake<C::Connected, B>)
+}
+
+pub struct ResponseFuture(conn::ResponseFuture);
+
+#[derive(Debug)]
+pub enum ConnectError<E> {
+    Connect(E),
+    Handshake(Error),
+}
+
+// ===== impl Connect =====
+
+impl<C, B> Connect<C, B> {
+    pub fn new<E>(connect: C, executor: E) -> Self
+    where
+        E: Executor<BoxSendFuture> + Clone + Send + Sync + 'static,
+    {
+        Connect {
+            connect,
+            executor: ArcExecutor::new(executor),
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<C, B> svc::Service<()> for Connect<C, B>
+where
+    C: connect::Connect,
+    C::Connected: Send + 'static,
+    B: Payload,
+{
+    type Response = Connection<B>;
+    type Error = ConnectError<C::Error>;
+    type Future = ConnectFuture<C, B>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        Ok(().into())
+    }
+
+    fn call(&mut self, _target: ()) -> Self::Future {
+        ConnectFuture {
+            executor: self.executor.clone(),
+            state: ConnectState::Connect(self.connect.connect()),
+        }
+    }
+}
+
+// ===== impl ConnectFuture =====
+
+impl<C, B> Future for ConnectFuture<C, B>
+where
+    C: connect::Connect,
+    C::Connected: Send + 'static,
+    B: Payload,
+{
+    type Item = Connection<B>;
+    type Error = ConnectError<C::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            let io = match self.state {
+                ConnectState::Connect(ref mut fut) => {
+                    try_ready!(fut.poll().map_err(ConnectError::Connect))
+                },
+                ConnectState::Handshake(ref mut fut) => {
+                    let (tx, conn) = try_ready!(fut.poll().map_err(|err| ConnectError::Handshake(err.into())));
+                    self
+                        .executor
+                        .execute(conn.map_err(|err| debug!("http2 conn error: {}", err)))
+                        .unwrap();
+
+                    return Ok(Connection { tx }.into());
+                }
+            };
+
+            let _e = &self.executor;
+            let handshake = conn::Builder::new()
+                .http2_only(true)
```

Comment:
does this disable push-promise by default?

Reply:
Yes: https://github.com/hyperium/hyper/blob/master/src/proto/h2/client.rs#L43